### PR TITLE
Store attendance in an attendees table

### LINKED ISSUE #167: Store attendance for an event
We currently don't know who attends an event since we store it only on the user side.

### PR DESCRIPTION
**Pull request by psrpinto**
Fixes #167 

I would recommend reviewing the final result instead of individual commits, as some iteration was required.

## Summary of changes

- Extract all `require_once` calls to an `autoload.php` file
    - And load it early instead of in `gp_init()`, to guarantee required files are always loaded
- Extract database upgrade logic to a separate file
- Move `Attendee_Repository` to an `Attendee` namespace
- Create `Attendee` class which represents a user attending an event
- Adapt `Attendee_Repository` so it uses `Attendee` class
- Add missing methods in `Attendee_Repository` (e.g. `get_attendee()`)
- Create `attendees` table, with the following fields
    - `translate_event_attendees_id` (primary key)
    - `event_id`: `int(10)`
    - `user_id`: `int(10)`
    - `is_host`: `tinyint(1)` (bool)

### CODE REVIEW

File: wporg-gp-translation-events.php
Comment (psrpinto) [inline wporg-gp-translation-events.php]: Should the indexes be namesapaced to translation-events, or is this fine?

File: wporg-gp-translation-events.php
Comment (psrpinto) [inline wporg-gp-translation-events.php]: The primary key is technically not needed but since the actions table above has it (and it's also not technically needed), I guessed I should probably add it here too.

File: wporg-gp-translation-events.php
Comment (psrpinto) [inline wporg-gp-translation-events.php]: This doesn't seem to be working somehow. Before spending time debugging it, I figured I should first confirm that we want to do this. We can always leave the meta behind, it won't have any impacts.

Let me know how to proceed.

File: wporg-gp-translation-events.php
Comment (akirk) [reply wporg-gp-translation-events.php]: Index names are just informative, I guess you'd only the name when you'd use `USE INDEX` which is a bit of a bad sign of too much table complexity.

**Comment by akirk** [comment]
Since we need a database upgrade, we need to be careful with errors. What if we added upgrade functionality like GlotPress with a version number? https://github.com/GlotPress/GlotPress/blob/develop/gp-settings.php#L137-L142

Otherwise what is your recommendation for the upgrade? Deactivate the plugin, update the code, activate it again?

File: includes/attendee-repository.php
Comment (akirk) [inline includes/attendee-repository.php]: Could we use `$wpdb->delete()`? Then we don't need all the above disables.

File: includes/attendee-repository.php
Comment (akirk) [inline includes/attendee-repository.php]: Why don't we use `$wpdb->insert()` here? Then we wouldn't need the disables from above.

File: includes/attendee-repository.php
Comment (akirk) [inline includes/attendee-repository.php]: Could you check again if these are needed?

File: includes/attendee-repository.php
Comment (akirk) [inline includes/attendee-repository.php]: Also here I am not sure why we need these, does phpcs really complain?

File: tests/attendee-repository.php
Comment (akirk) [inline tests/attendee-repository.php]: Check these also please.

File: wporg-gp-translation-events.php
Comment (akirk) [reply wporg-gp-translation-events.php]: Let's leave the meta behind and remove this.

**Review by akirk — COMMENTED** [review]
Functionally this works well. I wonder about all of the phpcs:disables and whether we really need them. I'd try to avoid them if possible and use `wpdb` functions.

**Comment by akirk** [comment]
I've been thinking about this. Should we rename the table also since our rename plan from attendees to RSVP? #176

**Comment by psrpinto** [comment]
I agree with all your remarks/comments, will update accordingly.

> Otherwise what is your recommendation for the upgrade? Deactivate the plugin, update the code, activate it again?
> What if we added upgrade functionality like GlotPress with a version number? 

Indeed I guess we would need to do something like that. I will look into the GlotPress's upgrade functionality with version number as you suggested.

File: includes/attendee-repository.php
Comment (psrpinto) [reply includes/attendee-repository.php]: We're now using `$wpdb->delete()`. It allows for removing `WordPress.DB.PreparedSQL.InterpolatedNotPrepared` but not the other ones.

File: includes/attendee-repository.php
Comment (psrpinto) [reply includes/attendee-repository.php]: Since we're doing an `insert ignore`, I looked for a way to do insert ignore with `$wpdb->insert()` but it doesn't seem to be possible. Should I leave as is, or should we first delete and then `$wpdb->insert()`?

Comment (psrpinto) [reply includes/attendee-repository.php]: Or we could throw an exception if we're trying to insert an attendee that already exists.

**Comment by akirk** [comment]
I don't recall where I read @amieiro saying this but I agree that we should not rename the table to rsvp. But I do think we should introduce the status fields as an enum with `rsvp`, `contributed`, `cohost` and we'll keep it at `rsvp` in this iteration.

**Comment by amieiro** [comment]
@akirk I commented this [here](https://github.com/WordPress/wporg-gp-translation-events/issues/176#issuecomment-2011783729).

File: includes/attendee-repository.php
Comment (psrpinto) [reply includes/attendee-repository.php]: I have confirmed that all these disables are needed, also the ones you mentioned in other comments.

The `InterpolatedNotPrepared` is because of the table name interpolation.

File: includes/attendee-repository.php
Comment (akirk) [reply includes/attendee-repository.php]: An `INSERT IGNORE` is just an insert that doesn't return an error if it couldn't insert a row because it already existed. So if we just ignore the return value then just using `$wpdb->insert()` should be fine?

File: wporg-gp-translation-events.php
Comment (amieiro) [inline wporg-gp-translation-events.php]: If you want to run this once in production and to avoid running on the tests, I think you need to invert the check:

```suggestion
			if ( 'true' !== getenv( 'TRANSLATION_EVENTS_NO_IMPORT_LEGACY_ATTENDEES' ) ) {
```

File: wporg-gp-translation-events.php
Comment (amieiro) [inline wporg-gp-translation-events.php]: The `get_contributors` method receive and `event id`, so you need to update the parameter passed.

```suggestion
			foreach ( $stats_calculator->get_contributors( $event->ID ) as $users ) {
```
This method returns a list of users, so I think you need to loop though the users and add all to the new table.

**Review by amieiro — CHANGES_REQUESTED** [review]

File: wporg-gp-translation-events.php
Comment (amieiro) [inline wporg-gp-translation-events.php]: Are you going to add here the new field for the user status in this event: attendee (default value), co-host,...?

**Comment by psrpinto** [comment]
Thanks for reviewing @amieiro, I'm picking up this PR again and will address your remarks.

**Comment by psrpinto** [comment]
I've rebased this with #191 so this PR now depends on that one. No other changes have been done, only rebased. New commits will be added after this comment, prior commits can be considered reviewed.

File: wporg-gp-translation-events.php
Comment (psrpinto) [reply wporg-gp-translation-events.php]: > The `get_contributors` method receive and event id, so you need to update the parameter passed.

Great catch, thank you! Fixed.

Comment (psrpinto) [reply wporg-gp-translation-events.php]: > This method returns a list of users, so I think you need to loop though the users and add all to the new table.

The `foreach` iterates over the result of the call to `get_contributors()`, and assigns the value at that index to the `$user` variable. I'm guessing you missed the `foreach`? Or I might be misunderstanding something.

File: wporg-gp-translation-events.php
Comment (psrpinto) [reply wporg-gp-translation-events.php]: Good point, I've added this to the TODO in the PR description.

File: wporg-gp-translation-events.php
Comment (psrpinto) [reply wporg-gp-translation-events.php]: Nice catch, fixed

File: includes/attendee-repository.php
Comment (psrpinto) [reply includes/attendee-repository.php]: I tried it and indeed it works, though warnings are written to the log, which also happens during tests, so it pollutes the test output. An example of these warnings:

> WordPress database error Duplicate entry '35-17' for key 'event_per_user'

Would it be ok to leave it as is?

**Comment by psrpinto** [comment]
This is now ready for re-review. I've updated the PR description with a summary of all changes introduced here.

Concerning tracking who is host/contributor, finally I went with two booleans instead of an enum, I believe that is an appropriate solution here. Let me know if you have some remarks about that, or about anything else.

I also opened #195 and #196 which I would tackle as follow-ups to this PR.

File: includes/database.php
Comment (amieiro) [inline includes/database.php:35]: The main host is stored in the `wp_posts` table as `post_author`, so maybe it is better to call the other host `cohosts`, so we can differentiate the main creator and the other hosts.

```suggestion
			`is_cohost` tinyint(1) default 0 not null comment 'Whether the user is a host of the event',
```

File: includes/database.php
Comment (amieiro) [inline includes/database.php]: Why we need this field, as we store all the translations or suggestions in the `wp_gp_event_actions` table, so we don't need this field? In fact, we can have conflicting information if we store it in both places.

File: includes/database.php
Comment (amieiro) [inline includes/database.php:64]: This method is not importing the data for the `is_host` and `is_contributor` fields.

File: includes/database.php
Comment (amieiro) [inline includes/database.php:32]: Why we are using the table name in the field name and not only the `id` word? If we change it, we should change the `PRIMARY KEY` field.

```suggestion
			`id` int(11) unsigned NOT NULL AUTO_INCREMENT,
```

File: includes/database.php
Comment (psrpinto) [reply includes/database.php]: I guess we could indeed calculate it from the `actions` table, though I don't know if that could be an expensive query, given the `actions` table will likely become a very large table with time.

For example, when displaying the event details page, we would need to calculate it for each attendee of the event.

Concerning conflicting information, I think that would not happen, because you can't go back to not being a contributor. I mean, if you contributed to an event, it's not possible to "undo" that contribution, right?

So once `is_contributor` is set to `true`, it will always be `true`, I think there is no scenario where it would become `false` again.

Personally I like that the data of attendees would be decoupled from the `actions`. If you think we should change it, I can do it.

File: includes/database.php
Comment (psrpinto) [reply includes/database.php:64]: ~I pushed a commit and the method is now setting `is_host`. Concerning, `is_contributor`, I would take care of setting it when I tackle #196. Would that be ok?~

File: includes/database.php
Comment (psrpinto) [reply includes/database.php:32]: This was done for consistency with the `event_actions` table, where the field is named `translate_event_actions_id`. I believe we named the primary key like that after feedback we received from wordpress.org maintainers, so I figured it would make sense to follow the same naming pattern here.

In fact, the primary key is technically not needed, but since the `event_actions` table has it (and it's also not technically needed), I guessed I should probably add it here too.

File: includes/database.php
Comment (psrpinto) [reply includes/database.php:35]: The idea would be that the post author is no longer used to determine whether a user is a host. Instead, the first `host` is set to be the post author when the event is published.

I think the advantage of doing it like this is that there's a single way to determine whether a user is a host. An added advantage, is that author and host become decoupled, so it's possible for an event to have a single host that is not the post author. This makes it possible to "transfer" host if, for example, the user who created the event can no longer attend.

I realise this isn't a feature we have now, just using it as an example of an advantage of decoupling author from host.

File: includes/database.php
Comment (amieiro) [reply includes/database.php]: My main concerns here are:
- If we have a bug in the code, we could have inconsistent information in the tables. E.g., translations for a user in the `event_actions` table and this user with `is_contributor=0` in the `event_ateendees` table. I could assume that this is unlikely.
- If we store this information in the `event_ateendees` table, each time a user makes a translation, we are going to need to query this table to know if this user is contributing to this event, so we need to do a lot of extra queries when the users translate strings. If we have this information in the `event_actions` table, we will only query it when we need to show info for an event, and the query is not expensive, because it only uses one table (will maybe a lot of registers). We could cache this query in the future. And as Alex suggested [here](https://github.com/WordPress/wporg-gp-translation-events/issues/196#issuecomment-2043171940), we could compute it once a day, with a cron, for the events that have finished the last day.

What do you think about?

Comment (psrpinto) [reply includes/database.php]: These are fair points 👍

I also thought further about this and another problem with having `is_contributor` in the attendees table is that if the user would un-attend the event, the `is_contributor` would be lost. So we would need to prevent attendees who contributed from un-attending, which is not something I'm certain we want to do.

I will change it to remove `is_contributor` from the attendees table.

Comment (akirk) [reply includes/database.php]: What about we call it `was_contributor` so that it's clear that this is supposed to be calculated after the event is over?

Comment (akirk) [reply includes/database.php]: The idea is that we can use the `event_attendees` table to generate stats without having to query the `event_actions` table.

Comment (psrpinto) [reply includes/database.php]: Given that these stats are not a feature we have right now, could we delay this decision until we implement those stats? At that point we can decide to introduce the `was_contributor` field, or to calculate those stats from both the `attendees` and `actions` tables.

Comment (akirk) [reply includes/database.php]: Ok! Better than delaying this more.

Comment (amieiro) [reply includes/database.php]: Ok, we can remove the `is_contributor` and move on with this PR to approve it.

Comment (psrpinto) [reply includes/database.php]: I'm on this.

**Comment by psrpinto** [comment]
I believe this is now ready. #195 and #199 will be follow-ups to this PR.

**Comment by psrpinto** [comment]
~Don't merge yet, I accidentally pushed a commit to another branch~

**Comment by psrpinto** [comment]
Now it's ready :)

**Comment by akirk** [comment]
Sorry, I am not sure I understand 7e5dd652909e216cf92615cd6aaa5cf611d9fe9a. Why do contributors of previous projects get added as host?

**Comment by psrpinto** [comment]
> Sorry, I am not sure I understand https://github.com/WordPress/wporg-gp-translation-events/commit/7e5dd652909e216cf92615cd6aaa5cf611d9fe9a. Why do contributors of previous projects get added as host?

Thanks for catching that, I had a brain freeze there, it was not what I intended to do, I got confused between contributor and host. I will revert that commit. 

The intention was to add the post_author of the event as an attendee with `host` set to `true`. I will tackle that in #195 (which I'm working on now), if that's ok with you.

File: includes/database.php
Comment (psrpinto) [reply includes/database.php:64]: I will tackle setting `is_host` for legacy attendees in #195 .

**Review by amieiro — APPROVED** [review]

File: wporg-gp-translation-events.php
Comment (akirk) [inline wporg-gp-translation-events.php:215]: @psrpinto and I talked about this and that it should probably rather be `$user_id = $post->post_author`

Comment (amieiro) [reply wporg-gp-translation-events.php:215]: So, if:
- User1 creates a draft event.
- User2 publishes this event.
- The user1 will be set as attendee and as host.

It makes sense to me. I will update it in https://github.com/WordPress/wporg-gp-translation-events/pull/200.

Comment (akirk) [reply wporg-gp-translation-events.php:215]: @amieiro actually made exactly that change in https://github.com/WordPress/wporg-gp-translation-events/pull/200/files/ca6c64390702d6e46f4302aa5cd835e202b6b010..5bd799f4757f54dfd871ad80aa98b1a8ae412512

Comment (akirk) [reply wporg-gp-translation-events.php:215]: Thanks, @amieiro!

**Pull request merged**
